pkg/upload/js.go

```diff
@@ -0,0 +1,376 @@
+	}
+
+	bundlePath, err = resolveBundlePath(bundlePath, sourceMapPath, logger)
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
+
+	url := bundleUrl
+	if baseUrl != "" {
+		_, fileName := filepath.Split(bundlePath)
+		url = baseUrl + fileName
+		logger.Debug(fmt.Sprintf("Generated URL %s using the base URL %s", url, baseUrl))
+	}
+
+	uploadOptions, err := utils.BuildJsUploadOptions(apiKey, appVersion, url, projectRoot, overwrite)
+
+	if err != nil {
+		logger.Fatal(fmt.Sprintf("failed to build upload options: %s", err.Error()))
+	}
+
+	fileFieldData := make(map[string]server.FileField)
+	fileFieldData["sourceMap"] = sourceMapFile
+	if bundlePath != "" {
+		fileFieldData["minifiedFile"] = server.LocalFile(bundlePath)
+	}
+
+	err = server.ProcessFileRequest(endpoint+"/sourcemap", uploadOptions, fileFieldData, timeout, retries, sourceMapPath, dryRun, logger)
+
+	if err != nil {
+		logger.Fatal(fmt.Sprintf("encountered error when uploading js sourcemap: %s", err.Error()))
+	}
+}
+
+func ProcessJs(
+	apiKey string,
+	versionName string,
+	bundleUrl string,
+	baseUrl string,
+	sourceMapPath string,
+	bundlePath string,
+	projectRoot string,
+	Path utils.Paths,
+	endpoint string,
+	timeout int,
+	retries int,
+	overwrite bool,
+	dryRun bool,
+	logger log.Logger,
+) error {
+
+	for _, path := range Path {
+
+		outputPath := path
+
+		// Set a default value for projectRoot if it's not defined
+		projectRoot := resolveProjectRoot(projectRoot, path)
+		logger.Debug(fmt.Sprintf("Using project root %s", projectRoot))
+
+		appVersion := resolveVersion(versionName, path, logger)
+
+		// Check that the source map(s) exists and error out if it doesn't
+		sourceMapPaths, err := resolveSourceMapPaths(sourceMapPath, outputPath, logger)
+		if err != nil {
+			return err
+		}
+
+		// Check that we now have a source map path
+		if len(sourceMapPaths) == 0 {
+			return fmt.Errorf("could not find a source map, please specify the path by using --source-map")
+		}
+
+		// Ensure that the correct one of --bundle-url and --base-url is specified
+		isFile := utils.FileExists(sourceMapPath) || !utils.IsDir(outputPath)
+
+		if isFile && bundleUrl == "" {
+			return fmt.Errorf("`--bundle-url` must be set when uploading a file")
+		}
+		if isFile && baseUrl != "" {
+			return fmt.Errorf("`--base-url` must not be set when uploading a file")
+		}
+		if !isFile && baseUrl == "" {
+			return fmt.Errorf("`--base-url` must be set when uploading from a directory")
+		}
+		if !isFile && bundleUrl != "" {
+			return fmt.Errorf("`--bundle-url` must not be set when uploading from a directory")
+		}
+
+		// Add a slash if it is not already on the end of the base URL
+		if len(baseUrl) > 0 && baseUrl[len(baseUrl)-1] != '/' {
+			baseUrl += "/"
+		}
+
+		for _, sourceMapPath := range sourceMapPaths {
+			uploadSingleSourceMap(bundleUrl, baseUrl, bundlePath, sourceMapPath, apiKey, appVersion, projectRoot, endpoint, timeout, retries, overwrite, dryRun, logger)
```

Comment:
For consistency with other commands, we should pass any error up for potential handling higher up. The idea of the CLI is to provide commands as building blocks, so we want to allow an umbrella command to call this and potentially handle an error.

Where there are potentially multiple files (e.g. dSYMs), we should exit out of the loop if there's an error. By calling fatal error you're doing this now, but we'll want to check the error returned and return it up the callstack.